extensions/localPublish/src/index.ts

```diff
@@ -124,9 +122,19 @@ class LocalPublisher implements PublishPlugin<PublishConfig> {
         await this.stopBot(botId);
       }
       if (!port) {
-        // Portfinder is the stablest amongst npm libraries for finding ports. https://github.com/http-party/node-portfinder/issues/61. It does not support supplying an array of ports to pick from as we can have a race conidtion when starting multiple bots at the same time. As a result, getting the max port number out of the range and starting the range from the max.
-        const maxPort = max(map(LocalPublisher.runningBots, 'port')) ?? 3979;
```

Comment:
@boydc2014 How different is this solution from max(map(LocalPublisher.runningBots, 'port')). Essentially, instead of lastUsedPort it was using the LocalPublisher.runningBots to get all ports of bots that are running/starting. In this way it was avoiding the conflicts.

Reply:
@liweitian The core issue is lets say we have 3 bots [bot1, bot2, bot3]. 

1. Bot1 is started.  it runs await runtime.build(runtimePath, project, fullSettings, port); before setting the port of the bot in the LocalPublisher object (this.setBot does this). Can you update the LocalPublisher.runningBots[projectId].port immediately after portfinder finds the port rather than assigning it after the runtime.build. In this way, the next bot that starts ignores that port as max(map(LocalPublisher.runningBots, 'port')) + 1 would be the new port.

Reply:
The difference is max(max...) runs before the port been find. For example, two co-routines both get the maxPort = 3980, and both looking for that, and both get a "pass" and both trying to start that port, and ops conflict happened. 

While the approach here is to let each coroutine find port first (which is async and can return same result), ok, after finding the target port, check again, if this port is already been used. Because the post part of the port finding is sync by the nature of node js, so only co-routine will declare it gets the port, and the other co-routine will have to find port again. That's how the conflict can be avoided.

Reply:
I made some changes on the code. After testing it again, the find port operation does not causes a race condition. It is probably not an multi-threaded IO operation but the runtime.build or some code after that causes the race condition. If we set the running bots' port right after the find port method. The conflict can be solved. And only setting the port without changing status won't cause any error on the FE. So it is better to use it to replace the lastUsedPort.

Reply:
I'm guessing, maybe actually the async operation is always correct, but it's just mean the async part in node is done, probably doesn't count the os part (which is the port eventually get used).

Reply:
![image](https://user-images.githubusercontent.com/3829387/123972984-ed800a00-d9ed-11eb-8e78-73ebe9b2092e.png)

I think it's the upper layer that is not waiting, so our solution is probably still have issues

Reply:
@boydc2014 Good catch. What if we just wrap the code to find a port which is currently inside publishAsync line 119 to 132. We do not ever want to await on runtime.build. But we can await until we find the port. In this way, Localpublisher.runningBots[projectId].port is set before the next bot is called

```javascript
publish = () => {
const port = await getPort()
 this.publishAsync(port)
}

const getPort = async () => // await portfinder.find() returns the port
```
Thoughts? @boydc2014 @liweitian

Reply:
That sounds good.